### PR TITLE
[SPARK-46819][CORE] Move error categories and states into JSON

### PR DESCRIPTION
**Pull request by nchammas**
### What changes were proposed in this pull request?

Move the error categories and states out of the Markdown table in the error README and into standalone JSON files.

I am using the terms "error category" and "error state" as defined in SPARK-46810.

### Why are the changes needed?

The data captured in Markdown is not automation-friendly. In future work, I intend to automate the generation of [sql-error-conditions-sqlstates.md][states] using these new JSON files.

Additionally, there are many instances of duplicate keys and inconsistent descriptions for the error categories, which this change resolves by moving the data into JSON. This also enables IDE support for basic validation.

[states]: https://github.com/apache/spark/blob/6ea2094da252ce96d875696dbd68c4d1bc673dc4/docs/sql-error-conditions-sqlstates.md

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

The updated unit tests ensure the following:
- All error categories (e.g. `42`) and states (e.g. `42K01`) have the correct length and format.
- All error states have a two-character prefix that is a valid error category.
- All error classes define an error state that is valid.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by nchammas** [comment]
@MaxGekk - Before I continue down this path, do you like where this is going?

Asking you per @itholic's [comment on SPARK-46810][1].

[1]: https://issues.apache.org/jira/browse/SPARK-46810?focusedCommentId=17810156&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17810156

**Comment by itholic** [comment]
also cc @cloud-fan

**Review by MaxGekk — COMMENTED** [review]
SGTM

**Comment by nchammas** [comment]
@MaxGekk - The naming choices I made here are based on my proposal in SPARK-46810. Are you OK with what I wrote there? (If yes, could you chime in on the ticket just for the record, please?)

**Comment by MaxGekk** [comment]
@nchammas I will review this PR tomorrow. Also cc @srielau

**Comment by nchammas** [comment]
No rush at all. (Didn't mean to nag. Dunno if my comment came off that way.)

**Comment by srielau** [comment]
Thanks for doing this. Adding a row to this table is royal pain in the behind. This will make it easier.

**Review by MaxGekk — COMMENTED** [review]
@nchammas I wonder did you generate [error-states.json](https://github.com/apache/spark/pull/44863/files#diff-09927682f352e689f38cdbce800dba011d33000d09b1b2c4ff9bd4aa49ac265f) automatically by a script?

**Comment by nchammas** [comment]
> @nchammas I wonder did you generate [error-states.json](https://github.com/apache/spark/pull/44863/files#diff-09927682f352e689f38cdbce800dba011d33000d09b1b2c4ff9bd4aa49ac265f) automatically by a script?

A little scripting and a little IDE kung-fu.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

File: common/utils/src/main/resources/error/README.md [hunk 1]
```diff
@@ -116,1241 +116,4 @@ The following SQLSTATEs are collated from:
 - PostgreSQL 15
 - Oracle 12 (last published)
 - SQL Server
-- Redshift.
-
-<!-- SQLSTATE table start -->
-|SQLSTATE |Class|Condition                                         |Subclass|Subcondition                                                |Origin         |Standard|Used By                                                                     |
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/README.md:122]: This is not public doc, correct?

**Comment by MaxGekk** [comment]
The failure [Run / Run Spark on Kubernetes Integration test](https://github.com/nchammas/spark/actions/runs/7664194822/job/20888189765#logs) is not related to the changes, I do believe.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @nchammas and @itholic @srielau for review.

File: common/utils/src/main/resources/error/README.md (hunk 1 above)
Comment (nchammas) [reply common/utils/src/main/resources/error/README.md:122]: Correct, not public. I will be opening separate PRs to update our public facing documentation on error classes.

**Pull request closed without merging**